rep-0153.rst

```diff
@@ -0,0 +1,94 @@
+REP: 153
+Title: ROS distribution files
+Author: Dirk Thomas <dthomas@osrfoundation.org>
+Status: Final
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 27-Oct-2018
+Post-History: 27-Oct-2018
+
+
+.. contents::
+
+Abstract
+========
+This REP updates the specification of the ROS distribution files facilitated in
+the building, packaging, testing and documenting process.
+
+The intention is to annotate ROS version with their major version (``1`` or
+``2``) and avoid the need to externally define and update a classification of
+ROS distribution names as ROS 1 or ROS 2.
```

Comment:
It seems to me that all that is needed to be specified here is the toolchain used for the given distro (e.g. catkin vs colcon?). It says so here https://github.com/ros2/ros_buildfarm_config/pull/12 ("The ros_version is necessary to decide about the installation of the catkin or ros-workspace package without the need to hardcode ROS distro names in the ros_buildfarm code")
This could be added to the REP as an example for the usage, btw.

One can imagine the possibility of distributions from 2 ROS major versions sharing the same toolchain one day. And more easily one can imagine a situation where distributions from the same ROS major version use two different toolchains (such as rosbuild and catkin for ROS1, exagerrating the technical issues only slightly).
So merely storing the ROS major version as an integer seems limiting, whereas storing the toolchain name seems to solve the problem at hand sufficiently and transparently.

Even if you have other resons to want to store ROS version numbers in the ROS distribution files, you still have to justify why not using the full semantic versioning, or version strings (given whatever purpose might arise in the future might have to discriminate between minor versions). Integers make for poor identifiers.

Apart from that, using 'ros_' as a prefix for a key tightly binds rosdistro as a tool to the ROS ecosystem and the ROS brand, which is slightly against the philosophy of creating flexible open-source tools that can be re-used for various projects (however unlikely that may seem right now).

If you still persist that you really, really just need the ROS major version as an integer, at least name the key "major_version".

Reply:
> Even if you have other resons to want to store ROS version numbers in the ROS distribution files, you still have to justify why not using the full semantic versioning, or version strings (given whatever purpose might arise in the future might have to discriminate between minor versions).

@tkruse "Semantic versioning" of ROS distros? That sounds bizarre to me since the definition of semver would imply that every distro must be a new major version (since at least something will be 
a breaking change) which would make all other parts of the version irrelevant. Imo a versioning scheme across such a large number of federated packages doesn't make any sense. Please feel free to propose something if you think otherwise. Atm I don't foresee a use case not satisfied by a single number. If you have one in mind please bring it up and changing the type can certainly be considered.

> using 'ros_' as a prefix for a key tightly binds rosdistro as a tool to the ROS ecosystem and the ROS brand, which is slightly against the philosophy of creating flexible open-source tools that can be re-used for various projects
> ...
> ... at least name the key "major_version".

I am not too worried about a "ros_" prefix in a database named `rosdistro` which is accessed by a Python package named `rosdistro`. The terminology of "major version" comes with a semantic meaning I am not sure is adequate in this case. Maybe there are other suggestions for naming the newly added key?

> Splitting the data concerning one distribution between index.yaml and distribution.yaml can cause further unnecessary headaches.

On distribution entry in the index file can have N distribution files. Therefore the version information should go into the index file. Otherwise nothing would prevent you from having multiple contradicting distribution files for a single distro.

Reply:
I meant to have a version string that not merely indicates the "major" ros version, but also contains an increasing number for each distribution (such that Melodic "1.13" is known to be the next after Lunar "1.12", even when the alphabet runs out after 'Z'). This can be a version string/list, or just having separate major_version and minor_version keys.

> The ros_version does not imply a 1-to-1 mapping to a build tool. The major version of a distro 
> is an "invariant". E.g. the need to install the catkin Debian package in order to have setup files 
> under /opt/ros/<distro> in ROS 1 or existence of a ros-<distro>-ros-workspace package in ROS2

Taking your example: with ROS2 Ardent and Bouncy, you need the existence of a ros-<distro>-ros-workspace package. But maybe with a future ROS2 distribution you would need something else. By relying on a ros_version, you establish a 1-to-1 mapping between the major version and something. Instead, it would be cleaner if there was no logic like "if ros_version == 1 then install catkin debian package". Rather, that decision in the code should depend on a different discriminator, which can either be explicit, or a version key that denominates the version of the prerequisites.

Explicit woud be like "prerequisites: ['catkin_debian_package']", a version key would be like "distribution_format: 'ROS1-catkin'". Using the ROS major version for anything else than display purposes establishes 1-on-1 relationships.

Reply:
Note my concerns are not functionally urgent, so no need to have a longer discussion about it.